src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -175,7 +183,7 @@ private static List<SemanticRange> CombineSemanticRanges(List<SemanticRange> ran
             var tokenModifiers = csharpResponse[i + 4];
 
             var semanticRange = CSharpDataToSemanticRange(lineDelta, charDelta, length, tokenType, tokenModifiers, previousSemanticRange);
-            if (_documentMappingService.TryMapToHostDocumentRange(generatedDocument, semanticRange.Range, out var originalRange))
+            if (_documentMappingService.TryMapToHostDocumentRange(generatedDocument, CopyValues(semanticRange, rangePool), out var originalRange))
```

Comment:
A new API here that operated on something other than `Range` would further reduce allocations, but seemed like a pretty big change for this late in the process. In future ideally we'll be able to move to auto-generated protocol types that aren't classes anyway, which will solve this too.

Reply:
Looked at this a bit more and its definitely a big change, just with how much churn it is in tests that use Moq 😁

Reply:
I take it back, wasn't too bad :)
https://github.com/dotnet/razor/pull/9285